beacon_node/network/src/sync/block_lookups/single_block_lookup.rs

```diff
@@ -343,336 +176,290 @@ impl<const MAX_ATTEMPTS: u8, T: BeaconChainTypes> SingleBlockLookup<MAX_ATTEMPTS
+    /// It has been checked for consistency between blobs and block, but no consensus checks have
+    /// been performed and no kzg verification has been performed.
+    Ok(RpcBlock<E>),
+    /// All child components have not yet been received.
+    DownloadIncomplete,
+    /// Child components should not be cached, send this directly for processing.
+    NotRequired,
+    /// There was an error during consistency checks between block and blobs.
+    Err(AvailabilityCheckError),
+}
+
+/// For requests triggered by an `UnknownBlockParent` or `UnknownBlobParent`, this struct
+/// is used to cache components as they are sent to the network service. We can't use the
+/// data availability cache currently because any blocks or blobs without parents
+/// won't pass validation and therefore won't make it into the cache.
+#[derive(Default)]
+pub struct CachedChildComponents<E: EthSpec> {
+    pub downloaded_block: Option<Arc<SignedBeaconBlock<E>>>,
+    pub downloaded_blobs: FixedBlobSidecarList<E>,
+}
+
+impl<E: EthSpec> From<RpcBlock<E>> for CachedChildComponents<E> {
+    fn from(value: RpcBlock<E>) -> Self {
+        let (block, blobs) = value.deconstruct();
+        let fixed_blobs = blobs.map(|blobs| {
+            FixedBlobSidecarList::from(blobs.into_iter().map(Some).collect::<Vec<_>>())
+        });
+        Self::new(Some(block), fixed_blobs)
     }
+}
 
-    pub fn downloading_peer(&self, response_type: ResponseType) -> Result<PeerShouldHave, ()> {
-        match response_type {
-            ResponseType::Block => self.block_request_state.state.peer(),
-            ResponseType::Blob => self.blob_request_state.state.peer(),
+impl<E: EthSpec> CachedChildComponents<E> {
+    pub fn new(
+        block: Option<Arc<SignedBeaconBlock<E>>>,
+        blobs: Option<FixedBlobSidecarList<E>>,
+    ) -> Self {
+        Self {
+            downloaded_block: block,
+            downloaded_blobs: blobs.unwrap_or_default(),
         }
     }
 
-    pub fn both_components_processed(&self) -> bool {
-        self.block_request_state.state.component_processed
-            && self.blob_request_state.state.component_processed
+    pub fn clear_blobs(&mut self) {
+        self.downloaded_blobs = FixedBlobSidecarList::default();
+    }
+
+    pub fn add_unknown_parent_block(&mut self, block: Arc<SignedBeaconBlock<E>>) {
+        self.downloaded_block = Some(block);
     }
 
-    pub fn set_component_processed(&mut self, response_type: ResponseType) {
-        match response_type {
-            ResponseType::Block => self.block_request_state.state.component_processed = true,
-            ResponseType::Blob => self.blob_request_state.state.component_processed = true,
+    pub fn add_unknown_parent_blobs(&mut self, blobs: FixedBlobSidecarList<E>) {
+        for (index, blob_opt) in self.downloaded_blobs.iter_mut().enumerate() {
+            if let Some(Some(downloaded_blob)) = blobs.get(index) {
```

Comment:
I don't see it specified in the blob lookup spec that the sidecars are ordered, is it safe to assume that peers return the blobs ordered by `blob_index`?

Reply:
Looks we only require order for by range responses. The `downloaded_blobs` here is always being constructed by us in lighthouse elsewhere though, so we should have inserted blobs based on blob index here everywhere it was constructed